### PR TITLE
Optimisation for 'delegate'

### PR DESCRIPTION
**Pull request by pjk25**
This is a manual rebase of rabbitmq-common #349

https://github.com/rabbitmq/rabbitmq-common/pull/349/commits/1c09c0f824cf2eca9284035db3f079465edb5733

Also see #2571

### CODE REVIEW

**Comment by pjk25** [comment]
@tomyouyou FYI

**Comment by pjk25** [comment]
This seems to have triggered some genuine failures in the `deps/rabbit/cluster_SUITE.erl`, in the `delegates_sync` testcase.

File: deps/rabbit_common/src/delegate.erl [hunk 1]
```diff
@@ -77,28 +88,47 @@ start_link(Name, Num) ->
...
+        {error, _, Error}  -> {error, Error}
+    end;
 invoke(Pid, Name, FunOrMFA) when is_pid(Pid) ->
     case invoke([Pid], Name, FunOrMFA) of
-        {[{Pid, Result}], []} ->
-            Result;
-        {[], [{Pid, {Class, Reason, StackTrace}}]} ->
-            erlang:raise(Class, Reason, StackTrace)
+        {[{Pid, Result}], []} -> Result;
+        {[], [{Pid, Error}]} -> {error, Error}
```
Comment (pjk25) [inline deps/rabbit_common/src/delegate.erl:109]: I suspect this change is the cause of the `cluster_SUITE` test failures

Comment (tomyouyou) [reply deps/rabbit_common/src/delegate.erl:109]: I remember that @michaelklishin has suggested such a change, and I forgot the specific reason.

**Comment by pjk25** [comment]
@mergifyio rebase

File: deps/rabbit_common/src/delegate.erl (hunk 1 above)
Comment (lukebakken) [reply deps/rabbit_common/src/delegate.erl:109]: This seems like a major change.

**Review by lukebakken — CHANGES_REQUESTED** [review]
This fails every time with these changes:

```
cd deps/rabbit
make ct-simple_ha t=cluster_size_3:auto_resume
```

**Comment by tomyouyou** [comment]
> This fails every time with the changes in the `rabbitmq-server-2571` branch:
> 
> ```
> cd deps/rabbit
> make ct-simple_ha t=cluster_size_3:auto_resume
> ```
@lukebakken
Do I need to close the PR and resubmit a new PR that contains the correction of the error?

**Comment by pjk25** [comment]
The `//deps/rabbit:cluster_SUITE` `delegates_sync` test also fails and is more tightly scoped to the change. I am having a look at it now. I don't think you would be able to push an additional commit onto this branch @tomyouyou , but feel free to open another and reference this one. For the moment we should leave this one open as well.

**Comment by pjk25** [comment]
Actually, I just realized I was not looking at the full diff for this PR, so rather than dig further, I will leave this to you @tomyouyou to open a new PR, and I will close this one.

**Comment by lukebakken** [comment]
> Do I need to close the PR and resubmit a new PR that contains the correction of the error?

Yes. Sorry for the inconvenience but when we merged repositories into a single one we couldn't migrate PRs. I opened this new one but due to how GH works you can't make changes.

If you'd like to continue working on this, open a new PR from your fork.

**Pull request closed without merging**